### DISCUSSION TITLE
Load .shp files into a Group/Node rather than a Map

### DISCUSSION THREAD

**Discussion by Warren-Whitfield** [pelicanmapping/osgearth#2563] [Support Q&A]
Is there a way to load .shp files into a group/node rather than the map? 
Currently they are loaded into the map as layers, but I rather want them in a group. Reason being, when I hide the map, the vector maps are obviously hidden as well because they are layers in the map.

**Comment by gwaldron** [comment]
Is this related to #2555? Are you hiding the MapNode or the TerrainEngineNode? If you just hide your TerrainEngineNode you should still see vector layers.

**Comment by Warren-Whitfield** [comment]
Yes it is related to #2555. I'm using the m_pVisibleCullingCallback->setVisible(false); as you suggested. It hides the entire map including all its layers. So I want to move the layers into their own group to overcome this.

**Comment by Warren-Whitfield** [comment]
So this is what I'm doing...
```
m_pVisibleCullingCallback = new osgEarth::Util::ToggleVisibleCullCallback();
m_pVisibleCullingCallback->setName("VisibleCullCallback object");
m_poMapNode->getTerrainEngine()->getNode()->addCullCallback(m_pVisibleCullingCallback );
```
...and it hides everything.

**Comment by gwaldron** [comment]
There are a couple approaches where vector data will render as terrain: (a) draping  (e.g., `altitude-clamping: terrain-drape;`), and (b) rasterized vector layers, i.e.  `<FeatureImage>` layers. If you are doing either of those, then hiding the terrain will also hide the vectors since they are actually drawn in conjunction with the terrain.

Anything you draw as either a `<FeatureModel>` layer or as Annotations will be OK.

**Comment by Warren-Whitfield** [comment]
So this is my function I'm currently using to load the files into the map. I'm not quite sure how to change it?

```
void cDisplayMap::addVectorData(Map* map, const std::string strName, const std::string url, osg::Vec4f color)
{
	//Create a feature source to load the shapefile
	OGRFeatureSource* data = new OGRFeatureSource();
	data->setURL(url);
	data->options().buildSpatialIndex() = true;

	//A resampling filter will ensure that the length of each segment falls within the specified range.
	//That can be helpful to avoid cropping very long lines segments.
	ResampleFilterOptions resample;
	resample.minLength() = 0.0f;
	resample.maxLength() = 25.0f;
	data->options().filters().emplace_back(resample);

	//Style
	Style style;
	style.setName(strName);

	//Render the data as lines that are 7.5m wide
	LineSymbol* line = style.getOrCreate<LineSymbol>();
	line->stroke()->color() = color;
	line->stroke()->smooth() = true;
	line->stroke()->width() = 1.0f;
	line->stroke()->roundingRatio() = 0.1f;
	line->tessellationSize()->set(100, Units::KILOMETERS);

	//Clamp the lines to the terrain
	AltitudeSymbol* alt = style.getOrCreate<AltitudeSymbol>();
	alt->clamping() = alt->CLAMP_TO_TERRAIN;

	//Apply a depth offset to avoid z-fighting. The "min bias" is the minimum
	//apparent offset (towards the camera) of the geometry from its actual position.
	//The value here was chosen empirically by tweaking the "oe_doff_min_bias" uniform.
	RenderSymbol* render = style.getOrCreate<RenderSymbol>();
	render->depthOffset()->minBias()->set(6.6, Units::METERS);

	//Create a model layer that will render the vectors according to our style sheet.
	osg::ref_ptr<FeatureImageLayer> layer = new FeatureImageLayer();
	layer->setName(strName);
	layer->setFeatureSource(data);
	layer->setStyleSheet(new StyleSheet());
	layer->getStyleSheet()->addStyle(style);
	layer->setOpenAutomatically(true);

	map->addLayer(layer);

	if (layer->getStatus().isError())
	{
		OE_WARN << layer->getStatus().message() << std::endl;
	}
}
```

**Reply by gwaldron** [reply]
Since you are using a `FeatureImageLayer` the nodemask approach won't work. You might need to install a custom layer shader instead. For example:
```c++
    const char* shader = R"(
        #pragma vp_function remove_background, fragment
        void remove_background(inout vec4 color) {
            if (distance(color.rgb, vec3(1)) < 0.1) discard;
        }
    )";
    auto vp = VirtualProgram::getOrCreate(layer->getOrCreateStateSet());
    ShaderLoader::load(vp, shader);
```

**Comment by Warren-Whitfield** [comment]
Hi Glenn, I'm not quite sure I understand how to use this? :thinking:

So I have the function above that loads the .shp files, then elsewhere I have that cullingCallback installed as you suggested in #2555. I then show/hide terrain with m_pVisibleCullingCallback->setVisible(true/false).

To toggle an arbitrary vector's visbility I do this...
```
osgEarth::LayerVector layers;
m_poMapNode->getMap()->getLayers(layers);
for (auto& layer : layers)
{
  FeatureImageLayer* fl = dynamic_cast<FeatureImageLayer*>(layer.get());
  if(layer->getName() == "RAILWAYS")
    fl->setVisible(true/false);
}
```
How do I now hide the terrain only and not the vector data using the custom shader you've shown?

My apologies for the ignorance, but this is all pretty new to me. (In fact, I had to AI the shader example above to get a basic understanding of what it does :rofl: )

**Comment by gwaldron** [comment]
What you're trying to do is not "officially" a feature in osgEarth, so I'm just giving you options and ideas.

Anything that's an "Image" layer (like `FeatureImageLayer`) is rasterized on to the terrain, and therefore you have to draw the terrain to see it, regardless of the fact that the images comes from vector data. The shader I posted will still draw the terrain but will mask out the white pixels "behind" the rasterized layer. It's a hack to make it work.

Things that just create OSG geometry, like `FeatureModelLayer`, are separate from the terrain and you can just turn off the terrain with the callback we talked about.

**Comment by Warren-Whitfield** [comment]
Thanks I do appreciate the effort. :+1: 
So for the white pixels being masked, is this assuming that the vector data is white so that it can be masked? Or am I missing something here?

**Reply by gwaldron** [reply]
No, the empty terrain geometry is white. So it's discarding any pixels that contain no data.
